pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go

```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return nil, fmt.Errorf("error getting signature bundle: %w", err)
+	}
+	sigContent, err := getBundleSignatureContent(bundle)
+	if err != nil {
+		return nil, fmt.Errorf("error getting signature content: %w", err)
+	}
+	selectorsFromSignatures.Content = sigContent
+
+	if bundle.Payload.LogID == "" {
+		return nil, fmt.Errorf("error getting signature log ID: %w", errors.New("empty log ID"))
+	}
+	selectorsFromSignatures.LogID = bundle.Payload.LogID
+
+	if bundle.Payload.IntegratedTime == 0 {
+		return nil, fmt.Errorf("error getting signature integrated time: %w", errors.New("integrated time is 0"))
+	}
+	selectorsFromSignatures.IntegratedTime = strconv.FormatInt(bundle.Payload.IntegratedTime, 10)
+	return selectorsFromSignatures, nil
+}
+
+// ShouldSkipImage checks the skip list for the image ID in the container status.
+// If the image ID is found in the skip list, it returns true.
+// If the image ID is not found in the skip list, it returns false.
+func (s *sigstoreImpl) ShouldSkipImage(imageID string) (bool, error) {
+	if len(s.skippedImages) == 0 {
+		return false, nil
+	}
+	if imageID == "" {
+		return false, errors.New("image ID is empty")
+	}
+	_, ok := s.skippedImages[imageID]
+	return ok, nil
+}
+
+// AddSkippedImage adds the image ID and selectors to the skip list.
+func (s *sigstoreImpl) AddSkippedImage(imageIDList []string) {
+	if s.skippedImages == nil {
+		s.skippedImages = make(map[string]bool)
+	}
+	for _, imageID := range imageIDList {
+		s.skippedImages[imageID] = true
+	}
+}
+
+// ClearSkipList clears the skip list.
+func (s *sigstoreImpl) ClearSkipList() {
+	s.skippedImages = nil
+}
+
+// ValidateImage validates if the image manifest hash matches the digest in the image reference
+func (s *sigstoreImpl) ValidateImage(ref name.Reference) (bool, error) {
+	dgst, ok := ref.(name.Digest)
+	if !ok {
+		return false, fmt.Errorf("reference %T is not a digest", ref)
+	}
+	desc, err := s.functionHooks.fetchImageManifestFunction(dgst)
+	if err != nil {
+		return false, err
+	}
+	if len(desc.Manifest) == 0 {
+		return false, errors.New("manifest is empty")
+	}
+	hash, _, err := v1.SHA256(bytes.NewReader(desc.Manifest))
+	if err != nil {
+		return false, err
+	}
+
+	return validateRefDigest(dgst, hash.String())
+}
+
+func (s *sigstoreImpl) AddAllowedSubject(subject string) {
+	if s.subjectAllowList == nil {
+		s.subjectAllowList = make(map[string]bool)
+	}
+	s.subjectAllowList[subject] = true
+}
+
+func (s *sigstoreImpl) ClearAllowedSubjects() {
+	s.subjectAllowList = nil
+}
+
+func (s *sigstoreImpl) EnableAllowSubjectList(flag bool) {
+	s.allowListEnabled = flag
+}
+
+func (s *sigstoreImpl) AttestContainerSignatures(ctx context.Context, status *corev1.ContainerStatus) ([]string, error) {
+	skip, err := s.ShouldSkipImage(status.ImageID)
+	if err != nil {
+		return nil, fmt.Errorf("failed attesting container signature: %w", err)
+	}
+	if skip {
+		return []string{signatureVerifiedSelector}, nil
```

Comment:
should another annotation also be added in the instance that an image is skipped.. Surfacing this information would be helpful from an audit perspective